Publish libms GitHub registry

lindot11 opened the pull request:
Added to lib-ms.yaml: 

- publish-package action to github registry
- yarn publish access restricted for test gitjhub action in own repo

Change in package.json

- changed username and version for test github action in own repo

lindot11 commented on .github/workflows/lib-ms.yml:
```diff
@@ -60,3 +60,34 @@ jobs:
           files: servers/lib/coverage/clover.xml
           flags: lib-microservice-tests
+
+  publish-package:
```
I did the job dependency on line 67 needs: test-lib-ms

lindot11 commented on .github/workflows/lib-ms.yml:
```diff
@@ -60,3 +60,34 @@ jobs:
+          cd servers/lib
+          yarn install
+          yarn build
+          yarn publish --access restricted
```
done

lindot11 commented on servers/lib/package.json:
```diff
@@ -1,6 +1,6 @@
 {
-  "name": "@dtaas/libms",
-  "version": "0.2.0",
+  "name": "@lindot11/libms",
```
done

lindot11 commented on servers/lib/package.json:
```diff
@@ -1,6 +1,6 @@
-  "name": "@dtaas/libms",
-  "version": "0.2.0",
+  "name": "@lindot11/libms",
+  "version": "0.2.9",
```
done

lindot11 commented on .github/workflows/lib-ms.yml:
```diff
@@ -60,3 +60,34 @@ jobs:
+    needs: test-lib-ms
+    permissions:
+      packages: write
+      contents: read
```
I added permissions before I tested with switching username.
Now I tested if permissions made a difference in r/w access to be able to publish package. Doesn't seem to make a difference, and can be removed while still able to publish package.

lindot11 commented on .github/workflows/lib-ms.yml:
```diff
@@ -60,3 +60,34 @@ jobs:
+    runs-on: ubuntu-latest
+    needs: test-lib-ms
+    permissions:
+      packages: write
```
I tested using another user account, forking the repo public and making a PR. The account was not able to complete the publish job as long as the account is not part of the organization. 

I tested several settings, by adding / omitting the permissions in package.json and in the Settings -> Actions -> General -> Workflow permission: set to restricted / permissive. 

Only if the account is a member of the organization it is able to complete the publish action.
Else I got the error:
 Couldn't publish package: "https://npm.pkg.github.com/@lindot11%2flibms: Permission installation not allowed to Create organization package"
info Visit https://yarnpkg.com/en/docs/cli/publish for documentation about this command.
Error: Process completed with exit code 1.

prasadtalasila replied:
Your [packages](https://github.com/lindot11?tab=packages&repo_name=DTaaS) shows correct information.
The problem I am referring to comes when I open a PR from [DTaaS-Public](https://github.com/prasadtalasila/DTaaS-Public) to this repo. My current understanding is that the package will be published with just the PR which is not desirable. The package needs to be published only when a commit happens in `feature/distributed-demo` branch. You can check this by creating a new repository say `lindot11/DTaaS-Publish` and opening a PR to `lindot11/DTaaS`.

lindot11 replied:
I added an condition to control the job execution only when pushed to the feature/distributed-demo branch.

`  publish-package:
    if: github.ref == 'refs/heads/feature/distributed-demo'
    name: Publish to GitHub Packages
    runs-on: ubuntu-latest
    needs: test-lib-ms`

I tested that it doesn't run when doing a PR  to the feature branch or when pushing from a different branch. 
When I pushed from my feature branch the job succeeded in publishing the package.

prasadtalasila commented:
@lindot11 few more suggestions to improve README.md:

1. Rename **Configuration Setup** to **Configure**
2. Please update the name of **User Commands** to just **Use**.  Please replace the existing content of this section with text from [npm packages](https://github.com/lindot11/DTaaS/blob/feature/distributed-demo/docs/developer/npm-packages.md#rocket-use-the-packages) page. Please do modify the name of package (@into-cps-association/libms) and registry name. New text to be added at the end of **Use** section:

    ```txt

    The microservice is available at: http://localhost:PORT/lib


    The [API](https://into-cps-association.github.io/DTaaS/development/user/servers/lib/LIB-MS.html) page shows sample queries and responses.

    ```
3. Move **Developer Commands**, **NPM Package** and **Service Endpoint** sections to new DEVELOPER.md file.

lindot11 commented:
I still need to improve README.md with the name of package (@into-cps-association/libms) and registry name.
But while I was checking if I could install from the new registry I got this error:

`lindot@lindot:~/Desktop/PRJ4_DTaaS/DTaaS/servers/lib$ sudo npm install --registry=https://npm.pkg.github.com @lindot11/libms@0.2.1`
`npm ERR! 404 Not Found - GET https://npm.pkg.github.com/@apollo%2fclient - npm package "client" does not exist under owner "apollo"`
`npm ERR! 404  '@apollo/client@3.8.6' is not in this registry.`

I think the @apollo/client in npmjs name is not equivalent to the one in the github registry that I think is called @apollographql/apollo-client

prasadtalasila commented:
> I still need to improve README.md with the name of package (@into-cps-association/libms) and registry name. But while I was checking if I could install from the new registry I got this error:
> 
> `lindot@lindot:~/Desktop/PRJ4_DTaaS/DTaaS/servers/lib$ sudo npm install --registry=https://npm.pkg.github.com @lindot11/libms@0.2.1` `npm ERR! 404 Not Found - GET https://npm.pkg.github.com/@apollo%2fclient - npm package "client" does not exist under owner "apollo"` `npm ERR! 404 '@apollo/client@3.8.6' is not in this registry.`
> 
> I think the @apollo/client in npmjs name is not equivalent to the one in the github registry that I think is called @apollographql/apollo-client

You need to use [scoped registries](https://docs.npmjs.com/cli/v10/using-npm/scope#associating-a-scope-with-a-registry)

prasadtalasila commented on servers/DEVELOPER.md:
This file should be in `servers/lib` directory.

prasadtalasila commented on servers/lib/README.md:
You added the npm package publish information here. It is already in the [docs](https://github.com/INTO-CPS-Association/DTaaS/blob/feature/distributed-demo/docs/developer/npm-packages.md). Only the information from README.md has to be moved to DEVELOPER.md. The link to npm-packages.md need to be updated as well.

prasadtalasila commented:
@lindot11 please add your name to contributors in package.json.

prasadtalasila commented:
I've made some changes to your PR and made commits on [PR-239](https://github.com/prasadtalasila/DTaaS-Public/tree/pr-239) branch. The markdown files have been updated.

With the workflow file you pushed, the package publishing works correctly. The package only gets published upon commits to `feature/distributed-demo` branch. The pull requests will not result in publishing of new packages. **So the workflow file is good**.

I've also made one change to the workflow file. The goal is to publish the package from two branches: `feature/distributed-demo` and `release-v*`.

Please see this [PR](https://github.com/prasadtalasila/DTaaS-Public/pull/150). In this case, the package gets published from push on pr-239 branch. See logs of [pr-239 push](https://github.com/prasadtalasila/DTaaS-Public/actions/runs/6753597220). The workflow also tries to publish in [attempted pull request checks](https://github.com/prasadtalasila/DTaaS-Public/actions/runs/6753597254). The only reason it fails is because commit on `pr-239` branch stole the march and published the same package ahead of time. This is unexpected behavior; do you know why?

lindot11 commented:
Fixed the push condition for the publish-package job and tested it with success. Looked trough the updated md files looks good. Added contributers to the package.json. 

Git issue should be done :)

prasadtalasila commented on servers/lib/package.json:
```diff
@@ -1,10 +1,12 @@
-  "name": "@dtaas/libms",
-  "version": "0.2.0",
+  "name": "@into-cps-association/libms",
+  "version": "0.2.1",
```
please bump the version to `0.3.1`

prasadtalasila commented on .github/workflows/lib-ms.yml:
```diff
@@ -60,3 +60,42 @@ jobs:
+
+    if: | 
+      github.event_name == 'push' &&
+      (startsWith(github.ref, 'refs/heads/feature/') || startsWith(github.ref, 'refs/heads/release-v'))
```
please provide link to your github action logs which publish package from a branch starting with `release-v*`.

lindot11 replied:
https://github.com/lindot11/DTaaS/actions/runs/6798405636

lindot11 commented:
added the registry option and merged the pr

prasadtalasila commented:
> added the registry option and merged the pr. action log: https://github.com/lindot11/DTaaS/actions/runs/6798898140

Thanks, but the publishing step fails. https://github.com/lindot11/DTaaS/actions/runs/6798898140/job/18484027169

lindot11 commented:
that was expected. I'm not part of the organization to publish for into-cps-association

prasadtalasila commented:
> > added the registry option and merged the pr. action log: https://github.com/lindot11/DTaaS/actions/runs/6798898140
> 
> Thanks, but the publishing step fails. https://github.com/lindot11/DTaaS/actions/runs/6798898140/job/18484027169

The publication step is verified. It works as expected.

prasadtalasila commented:
An error occurs when package install is attempted:

```sh
~/git/prasadtalasila/temp/nodejs/libms-add » npm install @lindot11/libms --registry=https://npm.pkg.github.com/                  npm ERR! code E401
npm ERR! 401 Unauthorized - GET https://npm.pkg.github.com/@lindot11%2flibms - authentication token not provided

npm ERR! A complete log of this run can be found in: /home/prasad/.npm/_logs/2023-11-08T15_45_39_896Z-debug-0.log

~/git/prasadtalasila/temp/nodejs/libms-add » npm install @prasadtalasila/libms --registry=https://npm.pkg.github.com/              npm ERR! code E401
npm ERR! 401 Unauthorized - GET https://npm.pkg.github.com/@prasadtalasila%2flibms - authentication token not provided

npm ERR! A complete log of this run can be found in: /home/prasad/.npm/_logs/2023-11-08T15_46_18_941Z-debug-0.log
```

lindot11 commented:
I updated the [README.md](https://github.com/lindot11/DTaaS/blob/publish-libms-github-registry/servers/lib/README.md#arrow_down-install) on how to install the package. You have to use your PAT to install the package.

I managed to install the package with these commands:
```bash
lindot@lindot:~$ npm login --auth-type=legacy --registry=https://npm.pkg.github.com
npm notice Log in on https://npm.pkg.github.com/
Username: lindot09
Password: Logged in on https://npm.pkg.github.com/.

lindot@lindot:~/Desktop/DTaaS/servers/lib$ sudo npm install -g --scope=@prasadtalasila/libms --registry=https://npm.pkg.github.com
npm verb cli /usr/bin/node /usr/local/bin/npm
npm info using npm@9.6.7
npm info using node@v18.18.2
npm verb title npm install
npm verb argv "install" "--global" "--scope" "@prasadtalasila/libms" "--registry" "https://npm.pkg.github.com"
npm verb logfile logs-max:10 dir:/root/.npm/_logs/2023-11-08T19_41_21_490Z-
npm verb logfile /root/.npm/_logs/2023-11-08T19_41_21_490Z-debug-0.log

changed 1 package in 1s
npm verb exit 0
npm info ok 
```

prasadtalasila commented:
> I updated the [README.md](https://github.com/lindot11/DTaaS/blob/publish-libms-github-registry/servers/lib/README.md#arrow_down-install) on how to install the package. You have to use your PAT to install the package.
> 
> I managed to install the package with these commands:
> ```bash
> lindot@lindot:~$ npm login --auth-type=legacy --registry=https://npm.pkg.github.com
> npm notice Log in on https://npm.pkg.github.com/
> Username: lindot09
> Password: Logged in on https://npm.pkg.github.com/.
> 
> lindot@lindot:~/Desktop/DTaaS/servers/lib$ sudo npm install -g --scope=@prasadtalasila/libms --registry=https://npm.pkg.github.com
> npm verb cli /usr/bin/node /usr/local/bin/npm
> npm info using npm@9.6.7
> npm info using node@v18.18.2
> npm verb title npm install
> npm verb argv "install" "--global" "--scope" "@prasadtalasila/libms" "--registry" "https://npm.pkg.github.com"
> npm verb logfile logs-max:10 dir:/root/.npm/_logs/2023-11-08T19_41_21_490Z-
> npm verb logfile /root/.npm/_logs/2023-11-08T19_41_21_490Z-debug-0.log
> 
> changed 1 package in 1s
> npm verb exit 0
> npm info ok 
> ```

It is weird. Why does a public package need authentication to install?
I am guessing that you just need to be a github user. Is that so?

lindot11 commented:
I looked around in the documentation, and it seems GitHub Packages requires authentication to ensure that only authorized users can access and install packages
[About scopes and permissions for package registries](https://docs.github.com/en/packages/learn-github-packages/about-permissions-for-github-packages#about-scopes-and-permissions-for-package-registries)

prasadtalasila commented:
Please mention that the password in the npm login is actually the github token.

There have been some recent pushes into `feature/distributed-demo` branch. It resulted in merge conflicts in README.md file. Please resolve the same.

The need to be in `servers/lib` to install `libms` is not helpful. We want users to install the package without having the source code of `servers/lib`. 
When we installed the package from `verdaccio`, we can install it without having source code. The npm packages also follow the same convention. For example, `sudo npm install -g serve`. How do we achieve this with github packages?

lindot11 commented:
Updated [README.md](https://github.com/lindot11/DTaaS/blob/publish-libms-github-registry/servers/lib/README.md#arrow_down-install). The corrected commands installs the packages without authentication and the source files.
